Optimizations for repeated VDS source file and dataset names

fortnern opened the pull request:
Detects when source file and dataset names are shared between multiple VDS mappings in the same layout, and only stores each once in memory, with other instances linking to the first mapping with the same name. Adds new file format for VDS mappings that only stores each repeated string once when beneficial and allowed by libver_bounds setting.

Resolves #5639 
<!-- ELLIPSIS_HIDDEN -->


----

> [!IMPORTANT]
> Optimizes VDS mappings by storing repeated source file and dataset names once, updating file format to 4.0, and adding tests for shared string functionality.
> 
>   - **Behavior**:
>     - Optimizes VDS mappings by storing repeated source file and dataset names once and linking others to it.
>     - Updates file format to 4.0 to support shared string storage, used when `libver_bounds` is 2.0 or later.
>   - **Functions**:
>     - Modifies `H5D__virtual_store_layout()` in `H5Dvirtual.c` to calculate and store shared strings.
>     - Updates `H5O__layout_decode()` in `H5Olayout.c` to decode shared strings.
>   - **Tests**:
>     - Adds `test_vds_shared_strings()` in `dsets.c` to verify shared string functionality.
>   - **Misc**:
>     - Updates `RELEASE.txt` to document file format change.
>     - Adds hash table adjustments in `uthash.h` for realloc scenarios.
> 
> <sup>This description was created by </sup>[<img alt="Ellipsis" src="https://img.shields.io/badge/Ellipsis-blue?color=175173">](https://www.ellipsis.dev?ref=HDFGroup%2Fhdf5&utm_source=github&utm_medium=referral)<sup> for aba9a006dd846b320a62a5963fb401a427a76c0d. You can [customize](https://app.ellipsis.dev/HDFGroup/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>


<!-- ELLIPSIS_HIDDEN -->

fortnern committed:
Add hash tables for detecting repeated source file and dataset names.
Share these repeated strings between mappings in memory.

github-actions[bot] committed:
Committing clang-format changes

fortnern committed:
Merge branch 'HDFGroup:develop' into vds_shared_strings

fortnern committed:
Spelling

mattjala commented on src/H5Pdcpl.c:
```diff
@@ -2080,6 +2082,10 @@ H5Pset_virtual(hid_t dcpl_id, hid_t vspace_id, const char *src_file_name, const
         size_t new_alloc = MAX(H5D_VIRTUAL_DEF_LIST_SIZE, virtual_layout.storage.u.virt.list_nalloc * 2);
 
+        /* Clear hash tables */
+        HASH_CLEAR(hh_source_file, virtual_layout.storage.u.virt.source_file_hash_table);
```
Why clear and rebuild the hash tables each time a mapping is added? I would think you could just check if the new mapping shares a file/dset name with any previous entry and add a key if it doesn't.

fortnern replied:
We need to clear the hash tables before we reallocate the array of entries, since the pointers in the hash table may no longer be valid after the realloc. If it proves to be a performance issue, we can add a feature to uthash that will update these pointers in place (unless it already has that feature and I missed it).

fortnern replied:
Note this clear doesn't happen every time we add a mapping, only when the list is realloced (it doubles every time we realloc).

fortnern committed:
Add new encoding format for VDS for shared names

github-actions[bot] committed:
Committing clang-format changes

fortnern committed:
Spelling

mattjala commented on src/H5Olayout.c:
```diff
@@ -575,10 +578,12 @@ H5O__layout_decode(H5F_t *f, H5O_t H5_ATTR_UNUSED *open_oh, unsigned H5_ATTR_UNU
-                        HGOTO_ERROR(H5E_OHDR, H5E_VERSION, NULL,
-                                    "bad version # of encoded VDS heap information, expected %u, got %u",
-                                    (unsigned)H5O_LAYOUT_VDS_GH_ENC_VERS, (unsigned)heap_vers);
+                    assert(H5O_LAYOUT_VDS_GH_ENC_VERS_0 == 0);
```
What's the purpose of this assert?

fortnern replied:
If the VERS_0 macro were not binary 0 (unsigned) we would need to check for heap_vers < VERS_0 in the range check below

mattjala replied:
Are you thinking about the possibility that we might want to add a `VERS_ERROR` value later, or something like that? Other than that, I can't think of a reason we'd expect VERS_0 to change, and asserting the value of fixed enum constants is pretty unusual.

fortnern replied:
I simply put that insert in there to document why we were only checking for the decoded version being outside of the range on the high side. Normally when you check for a number being out of range you have to check for both above the maximum and below the minimum. Checking for below the minimum here would be a no-op and would generate a warning about checking  for unsigned < 0 is always false.

mattjala committed:
Add test for VDS shared string behavior

mattjala committed:
Add re-open tests

brtnfld commented:
@ellipsis-dev, please review this pr

mattjala committed:
Initial introduction of new format spec

mattjala committed:
Add release note

mattjala commented on doxygen/dox/H5.format.4.0.dox:
```diff
@@ -0,0 +1,12706 @@
+</tr>
+</table>
+
+\subsection subsec_fmt3_infra_globalheapvds III.F. Disk Format: Level 1F - Global Heap Block for Virtual Datasets
```
This is the only section of this document that isn't directly copied from the 3.0 file spec

fortnern committed:
Fix potential bug with "corrupt" files. Comment changes.

fortnern committed:
Merge branch 'vds_shared_strings' of github.com:fortnern/hdf5 into vds_shared_strings

github-actions[bot] committed:
Committing clang-format changes

mattjala committed:
Update doxygen references

mattjala committed:
Add mixed/many-mapping test cases

mattjala committed:
Add mixed/many-mapping test cases

fortnern commented on test/dsets.c:
```diff
@@ -16148,108 +16152,1085 @@ test_dcpl_layout_caching(H5D_layout_t layout_type)
+    h5_fixname(FILENAME[0], fapl, filename, sizeof filename);
+
+    nerrors += (test_vds_shared_strings(fapl) < 0 ? 1 : 0);
+    exit(0);
```
I think this exit(0) needs to be removed

mattjala replied:
Yep, left that in from testing. Just fixed it

fortnern replied:
It looks like the exit(0) is still there. It looks like you used a force push instead of a regular commit to attempt to change it, and it must not have worked. I'm not really sure what that message means.

fortnern committed:
Implement in-place pointer offsets in uthash with HASH_ADJUST_PTRS
macro. Use this instead of clearing and rebuilding the hash tables when
doubling the size of the VDS mapping array.

fortnern committed:
Merge branch 'vds_shared_strings' of github.com:fortnern/hdf5 into vds_shared_strings

github-actions[bot] committed:
Committing clang-format changes

mattjala committed:
Remove debug short-circuit

fortnern commented:
Looks good to me. Thanks for your contributions Matt!

fortnern committed:
Rename buf_diff to ptr_adj in HASH_ADJUST_PTRS

brtnfld commented:
@ellipses-dev, review PR

fortnern committed:
Assert vds entries found from a hash table aren't above the highest valid
entry in the array of entries.

github-actions[bot] committed:
Committing clang-format changes
